irc/client.go

```diff
@@ -72,139 +72,144 @@ func (cc *clientConnection) finishRegistration() {
 }
 
 func (cc *clientConnection) handleConnInput() {
-	defer func() {
-		close(cc.shutdown)
-		cc.conn.Close()
-	}()
+	defer cc.conn.Close()
 
 	s := bufio.NewScanner(cc.conn)
 
-ScanLoop:
-	for s.Scan() {
-		// Reads a line with s.Text() and parses it as
-		// an IRC message
-		msgStr := s.Text()
-		msg, err := StringToMessage(msgStr)
-
-		// Pretty sure error handling shouldn't be this complicated
-		if err != nil {
-			if err == ErrMalformedIRCMessage {
-				log.Printf("[%v] sent malformed IRC message: %v", cc, msgStr)
-			} else if numeric, ok := err.(*NumericReply); ok {
-				fmt.Fprintln(cc.conn, cc.reply(*numeric).String())
-			} else {
-				log.Printf("[%v] error: %v", cc, err)
-			}
-			continue
-		}
+SelectLoop:
+	for {
+		select {
+		case <-cc.shutdown:
+			return
 
-		switch msg.Cmd {
-		case PrivmsgCmd:
-			// Swallow the PRIVMSG if we haven't registered yet
-			if cc.state != clientStateRegistered {
+		default:
+			if !s.Scan() {
```

Comment:
you should probably `close(cc.shutdown)` before `continue` so we can handle the EOF, right?